### ISSUE TITLE
bug: conflicting `ctx` types allowed at compile time

### ISSUE THREAD

**Issue by mmkal** [trpc/trpc#4306]
### Provide environment information

```
  System:
    OS: macOS 13.2.1
    CPU: (12) arm64 Apple M2 Max
    Memory: 57.83 GB / 96.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 18.15.0 - ~/.nvm/versions/node/v18.15.0/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v18.15.0/bin/yarn
    npm: 9.5.0 - ~/.nvm/versions/node/v18.15.0/bin/npm
  Browsers:
    Chrome: 112.0.5615.137
    Safari: 16.3
```

```
"@trpc/server": "10.23.0",
```

### Describe the bug

routers don't seem to check the types of their procedures or sub-routers. self contained example:

```ts
import {initTRPC} from '@trpc/server'
import {createNextApiHandler} from '@trpc/server/adapters/next'

const trpca = initTRPC.context<{a: string}>().create()
const trpcb = initTRPC.context<{b: string}>().create()

const parent = trpca.router({
  a: trpca.procedure.query(({ctx}) => ctx.a.slice()),
  b: trpcb.procedure.query(({ctx}) => ctx.b.slice()),
})

export default createNextApiHandler({
  router: parent,
  createContext: () => ({a: 'aaa'}),
})
```

this is fine at compile time, and hitting `/api/trpc.a` works, but hitting `/api/trpc.b` gets `TypeError: Cannot read properties of undefined (reading 'slice')` (since `ctx.b` is never defined anywhere) - I shouldn't have been allowed to use `trpcb.procedure.query(...)` on a router for `trpca`, since they have different contexts.

### Link to reproduction

https://www.typescriptlang.org/play?ts=5.0.3#code/JYWwDg9gTgLgBAb2AO2DAKgJQAoGEC+cAZlBCHAOQACMUYAxgPQDOAplAG7sUBQoksRPSisAhjFYA5VgA8YAQTDAAEqOQATADbtCJMpRp0mbTu0aj1osBKjNGyWTF496EZM3i0GouAF44KGhYeAB0rsgScgA8CKIAXHAeUCgA5vgAfAAUAJRhIuKsOS5uHnBe9ABGfgGoGDi4YW6RMDEVCUmpGTl5YhJFxe7wYKIiEdXloiGkAK42mQg8cHDxZUaTYKT0rOrTIiEAjtPsAJ6Z8-QwMvjZfulwFzIhk8yawFs52QA0i3BtqwwVEIbCBbHZ7Q4nM4IB7XW73S4hQEvN6FbJfHjXHg8WQCeDqVhEUTTTTwYS9KSORQqNRadjzH4zGwJYajGDfJZkgq4JqOBI5OHzFYUUQiijXb7XIA

### To reproduce

Paste the code above into a next api handler. I tried that above using the stackblitz suggested but there's a strange compile error that I don't get locally, it doesn't seem to acknowledge that the export for the next adapter exists: https://stackblitz.com/edit/github-qwqz1i-qs9elv?file=src/pages/api/trpc/[trpc].ts

### Additional information

My use-case: I'm trying to make a router in a library, which can be plugged into an application's parent router for modular functionality, with the requirement that the parent has a context which meets the requirement of the library router. But the type system doesn't seem to enforce this. Maybe it's just not supported (yet)?

I'm putting yes I'd be down to fix this but I don't know if I'll have the time to so if someone else wants to, go for it!

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

**Comment by Nick-Lucas** [comment]
> My use-case: I'm trying to make a router in a library, which can be plugged into an application's parent router for modular functionality, with the requirement that the parent has a context which meets the requirement of the library router. But the type system doesn't seem to enforce this. Maybe it's just not supported (yet)?

So i've tried _really hard_ to do something similar and I've concluded that tRPC's types are just not easy to achieve it with if it's even possible, I think we're actually about to abandon the idea entirely: https://github.com/trpc/trpc/pull/4200

We do have `.experimental_concat` which was another attempt (with its own issues) but I personally now highly recommend avoiding any pattern which involves creating or emulating a second `t` object and trying to use it together, because it's a minefield of issues. Instead write your library as a generic router factory which must be passed a baseProcedure with a generic type and then utilises it for stuff:  https://dev.to/nicklucas/trpc-patterns-router-factories-and-polymorphism-30b0 

If we fix the issue you're alluding to, it will probably be by **banning** the use of a separate `initTRPC.create()` against an existing one, possibly by attaching a unique symbol to the instance, which would entirely ban the combining of disparate `t` objects event if the shape is compatible.

**Comment by Nick-Lucas** [comment]
Just remembering that in my blog post I didn't write it with external libraries in mind, it imports `t.router` directly, the tweak for an external library is simple enough though, here's an example from our tests: https://github.com/trpc/trpc/blob/3d003ae17a5678d30e688afb92e4b9a59454f669/packages/tests/server/react/polymorphism.factory.tsx#L48

**Comment by mmkal** [comment]
Yeah, having the library expose a factory instead is what I'm leaning towards too. Still, I think this is a bug, so if it's very hard to fix, banning "t-mixing" seems like the good option for now - then it becomes a missing feature.

**Comment by Nick-Lucas** [comment]
Yep, definitely try and use a factory. I'm acutely aware of `HOC` syndrome reaching tRPC though. `withMMKALThing(withLogging(withAuth(t.procedure)))` is just not a beautiful thing, so if you have any questions or feedback as you build please bring it back to us. I'm really keen to provide better APIs for library authors :)

PRs very welcome on banning mixed `t` instances, so we'll keep this issue open

**Comment by mmkal** [comment]
Ah, actually that example (https://github.com/trpc/trpc/blob/3d003ae17a5678d30e688afb92e4b9a59454f669/packages/tests/server/react/polymorphism.factory.tsx#L48-L61) might not be possible for an external library, it doesn't seem like `createRouterFactory` is exposed. Is there an example of something external like the `withLogging` and `withAuth` HOCs above that you know of?

---

Another thing that occurred to me is that we night need to pass `t` rather than `t.procedure` into the factory - say if the factory wanted to create a subrouter. That could present another problem - I don't remember where the issue is, but as far as I'm aware context-swapping is only on the procedure level, not at the router or `t`-level. I originally implemented context-swapping in v9 at the router level, but I think somewhere in the v10 update it got changed to procedure-only. I think I remember @KATT saying it would stay that way unless there's a use-case for `t`-level context-swapping.

The use case could be: a library provides a router factor that requires a `t` with `context: { userId: string }`. The application has a `t` with `context: { userId: number }`. To use the library, you'd need to be able to use middleware on `t`, maybe something like:

```ts
import {createLibRouter} from 'some-public-library'

const t = initTRPC.context<{ userId: number }>()

export const appRouter = t.router({
  someAppThing: t.router({
    myProc: t.procedure.query(() => 'x'),
  }),
  // createLibRouter demands a `t` with context type `{userId: string}`
  someLibThing: createLibRouter(
    t.use(({ ctx, next }) => {
      return next({ ctx: { userId: ctx.userId.toString() } })
    }),
  ),
})
```

**Comment by Nick-Lucas** [comment]
> might not be possible for an external library, it doesn't seem like `createRouterFactory` is exposed. Is there an example of something external like the `withLogging` and `withAuth` HOCs above that you know of?

It doesn't need to be exposed, that's the userland code you need to copy & paste (and adapt) into your own codebase, the only novel thing here is the use of generics so that the output is compatible with the procedure being extended.

> Another thing that occurred to me is that we night need to pass `t` rather than `t.procedure` into the factory

This is also possible, just update the generics to take a `t` instance with something like 

```ts
// We might add this to tRPC, playing with it in another branch already
type AnyTRPCInstance = ReturnType<
  ReturnType<typeof createTRPCInner<any>>
>;

// your generic is:
// TInstance extends AnyTRPCInstance
```

> The use case could be: a library provides a router factor that requires a `t` with `context: { userId: string }`. The application has a `t` with `context: { userId: number }`. To use the library, you'd need to be able to use middleware on `t`, maybe something like:

I think you realistically want to be using Context Extension (we've moved away from calling it "Swapping" recently) on a base procedure anyway. Trying to manipulate the `t` object is unnecessary since middlewares can extend context freely and it's the only place context gets used

**Comment by mmkal** [comment]
> > might not be possible for an external library, it doesn't seem like `createRouterFactory` is exposed. Is there an example of something external like the `withLogging` and `withAuth` HOCs above that you know of?
> 
> It doesn't need to be exposed, that's the userland code you need to copy & paste (and adapt) into your own codebase, the only novel thing here is the use of generics so that the output is compatible with the procedure being extended.

I don't think this is right. Here's what happens if copy the imports in that file:

<img width="878" alt="image" src="https://user-images.githubusercontent.com/15040698/236250721-b5c1dec9-9dce-46fe-991c-8d43a202ef36.png">

Here's what happens if I change `@trpc/server/...` to `@trpc/server/dist/...` to make the compile errors go away:

<img width="1042" alt="image" src="https://user-images.githubusercontent.com/15040698/236250895-a1ac98b8-f151-48e1-881e-5cd658cd2dba.png">

Basically, as far as I understand, you're able to do that in the trpc repo because the trpc monorepo setup load modules slightly differently than a regular npm-installed package, which relies on the `exports` field in package.json. At least I think that's the reason it works in this repo, but either way the error message I'm getting seems pretty explicit - it needs to be exposed via `exports`.

> 
> > Another thing that occurred to me is that we night need to pass `t` rather than `t.procedure` into the factory
> 
> This is also possible, just update the generics to take a `t` instance with something like
> 
> ```ts
> // We might add this to tRPC, playing with it in another branch already
> type AnyTRPCInstance = ReturnType<
>   ReturnType<typeof createTRPCInner<any>>
> >;
> 
> // your generic is:
> // TInstance extends AnyTRPCInstance
> ```
> 
> > The use case could be: a library provides a router factor that requires a `t` with `context: { userId: string }`. The application has a `t` with `context: { userId: number }`. To use the library, you'd need to be able to use middleware on `t`, maybe something like:
> 
> I think you realistically want to be using Context Extension (we've moved away from calling it "Swapping" recently) on a base procedure anyway. Trying to manipulate the `t` object is unnecessary since middlewares can extend context freely and it's the only place context gets used

Ok - I still think there could be difficulty making the types work, if this bug is fixed, because the library might need to use `t.router` and `t` has a particular context type baked into it, but maybe that problem is small/easy to solve.

**Comment by Nick-Lucas** [comment]
Sorry, I did misunderstand slightly. So that piece is one of my first contributions to tRPC and I don't think you need the exact types I constructed there, they're actually overkill.

```ts
type RouterFactory<TConfig extends AnyRootConfig> = ReturnType<
  typeof createRouterFactory<TConfig>
>;
type BaseProcedure<TConfig extends AnyRootConfig> = ReturnType<
  typeof createBuilder<TConfig>
>;
```

Instead of the above, you'll want something like:

```ts
type BaseProcedure = ProcedureBuilder<any>
```

`RouterFactory` I think we might need to expose a new type for, but for the needs you described here you might not need that?

**Comment by Nick-Lucas** [comment]
> possibly by attaching a unique symbol to the instance

Just had a poke at this solution and it doesn't seem viable. We do stuff like `typeof unsetMarket` in the codebase by that's a statically known symbol. Dynamic symbols it looks like don't really work because almost no call-sites can verify that an incoming symbol is the right one statically. Maybe I'm wrong though, feel free to have a play with branch: `4306-experiment`

**Comment by KATT** [comment]
Actually, I don't think we will fix this.

We are planning to make sure the "built" procedures won't be aware of the router so we can decouple procedures from routers to try to reduce the noise of `.d.ts`-files and possibly improve perf (& having standalone procedures within RSC)

After calling `.query()`, we're planning for the resolved type only to include what is now `_input_in` and `_output_out`.

#4709

**Comment by mmkal** [comment]
Isn't this more about the router being aware of the procedures than the other way around?

In the original example, both procedures are perfectly valid. What wasn't valid, though, was putting the `trpcb` procedure in a router with `trpca`'s context. I think it's quite important that procedures report their required context. IMO the example in the OP is not unrealistic, I've hit it a few times.

To me it feels like the context is very important for the server implementation, so procedures and routers should know it somehow. But maybe there's a way to throw out the context when exposing the server API? Or am I misunderstanding this?

**Comment by KATT** [comment]
Yeah, the thinking is that the procedures themselves once baked are stupid input/output and you know more of the procedures by looking at the context.

It'll make for prettier types and autocompletion in the editor when hovering, smaller d.ts files and *hopefully* better perf (but if it isn't a perf improvement at all we might just drop it too and reconsider this)

Please join the discussion on the issue, I always value your feedback!

**Comment by Seven380** [comment]
Opened PR #7377 to address this by carrying procedure context types into router input validation.
